### ISSUE TITLE
Division and modulus operation definitions incorrectly depend on `Data.Nat.Properties`

### ISSUE THREAD

**Issue by MatthewDaggitt** [agda/agda-stdlib#1408]
Currently `_%_` and `_/_`  cannot be used without importing a ton of modules as they use the decidability proof over naturals in the proof of non-zeroness argument and therefore require `Data.Nat.Properties`.

In v2.0 after we swap to using the new `NonZero` predicate (see #921) we can move these definitions to `Data.Nat.Base`.

**Comment by mechvel** [comment]
When an user applies `x / y` or  `x % y`, most often one still needs a proof for `y /= 0`. 
For example, if `y` is originated as  ``y = a ∸ z``,  or as   ``y = a * z``,  the user still needs to compose a proof for `y /= 0`
in order to apply ``x / y``. And this effect does not depend on whether the NonZero  format is used or not. 
This is the most common case on practice. So that: once the user applies divMod, one most often has to use
Nat.Properties.

One could think of omitting the condition of `y /= 0` in the definition of divMod. This will preserve some properties of divMod, but break other important properties, so that some property formulations will need to add the condition ``y /= 0 -> ``.
Also this will need to change the name, because the classical names of ``/, mod, rem, divRem, divMod`` are already 
related to certain fixed properties. I do not think that this attempt is appropriate.

**Comment by MatthewDaggitt** [comment]
> For example, if y is originated as y = a ∸ z, or as y = a * z, the user still needs to compose a proof for y /= 0
in order to apply x / y. And this effect does not depend on whether the NonZero format is used or not.
This is the most common case on practice. So that: once the user applies divMod, one most often has to use
Nat.Properties.

In many cases the non-zero proof can be inferred automatically, such as in `n / suc d` or `7 / 3`. The former case occurs in `Data.Rational.Base`. So I'm afraid I disagree with you that you always need `Data.Nat.Properties`.

> One could think of omitting the condition of y /= 0 in the definition of divMod. ... I do not think that this attempt is appropriate.

Agreed

**Comment by mechvel** [comment]
Once the user applies divMod, one most likely is going to use the properties of divMod expressed not only via `/=`,
but also via ``+, *, == `` (say,  ``dividend == quotient * divisor + remainder``)   Nat.Properties.

Evoiding unnecessary large file import can be more naturally supported by arranging, for example: 
Nat/Properties.agda
Nat/Properties/Additive.agda
Nat/Properties/Multiplicative.agda
Nat/Properties/AddMulMix.agda 
Nat/Properties/Rest.agda,
where  Properties.agda  reexports all others.

This will help for many functions. But not for divMod. Because using divMod usually implies importing several of the above files,
which leads to importing Nat.Properties.

**Comment by mechvel** [comment]
> In many cases the non-zero proof can be inferred automatically, such as in n / suc d

Using `suc` for a proof of ``(/= 0#)`` is not so a good approach. It is ad hoc.
Because for Integer, `suc` does not help this; also for Rational and for many other domains the relation ``/= 0#`` has sense while the `suc` operation is senseless. The domain  `P = Rational[x]` of polynomials with rational coefficients has a sensible divMod
operation with the needed properties, and  divMod f g   relies on `g \= 0`, but `suc` s senseless for P.

**Comment by MatthewDaggitt** [comment]
> Once the user applies divMod, one most likely is going to use the properties of divMod expressed not only via /=,
but also via +, *, ==, _≤_ (say, dividend == quotient * divisor + remainder) . This causes import of Nat.Properties.

Not necessarily. If you are only performing a computation and not trying to prove anything about it then you won't need to import `Data.Nat.Properties`.

> If is better to follow an uniform expression for the laws of division with remainder,
and all these domains have a sensible relation of \= 0#.

Agreed, that's why in version 2.0 after implementing #921 we will be able to rewrite equation (I) as:
```
m≡m%n+[m/n]*n : ∀ m n {{.(NonZero n)}} → m ≡ m % n + (m / n) * n
```

**Comment by mechvel** [comment]
> If you are only performing a computation and not trying to prove anything about it then you won't need to 
> import Data.Nat.Properties.

Then, put  ` _/_` and `_%_` to Nat.Base.  And Nat.DivMod (or Nat.DivMod.Properties) could implement the corresponding proofs.
For the backwards compatibility, you can reexport  `_/_` and  `_%_` by Nat.DivMod.
What is the problem?

For example , ``x / 2``  probably does not need to care for proofs, and Nat.Base looks like sufficient.
If y is originated as ``y = <arithmetic expression>``, then probably Nat.Properties is need for the user.
So, if the user does not care for proofs, then in about 1/2 of cases  Nat.Properties still occurs needed.
Am I missing something?

**Comment by mechvel** [comment]
>> For the backwards compatibility, you can reexport `_/_` and `_%_` by Nat.DivMod.

But Nat.DivMod imports Nat.Properties.
Probably, you will need to sacrifise backwards compatibility in this case.
And as to me, I would remain `_/_` and `_%_` in Nat.DivMod.

**Comment by MatthewDaggitt** [comment]
> Then, put _/_ and _%_ to Nat.Base. And Nat.DivMod (or Nat.DivMod.Properties) could implement the corresponding proofs.
For the backwards compatibility, you can reexport _/_ and _%_ by Nat.DivMod.

Yes this is exactly the plan... but as I say at the start we can't do this until we switch over to using the new `NonZero` predicate!

**Comment by mechvel** [comment]
Probably you mean this:
``` 
NonZero zero    = ⊥
NonZero (suc x) = ⊤
```
I cannot tell right now what will be the whole consequences.

**Comment by MatthewDaggitt** [comment]
Fixed in #1697
